llvm/docs/Telemetry.rst

```diff
@@ -0,0 +1,246 @@
+  framework and any tool that wants to enable telemetry.
+* ``llvm::telemetry::TelemetryInfo``: Data courier
+* ``llvm::telemetry::Destination``: Data sink to which the Telemetry framework
+  sends data.
+  Its implementation is transparent to the framework.
+  It is up to the vendor to decide which pieces of data to forward and where
+  to forward them to for their final storage.
+* ``llvm::telemetry::Config``: Configurations for the ``Manager``.
+  
+.. image:: llvm_telemetry_design.png
+
+How to implement and interact with the API
+------------------------------------------
+
+To use Telemetry in your tool, you need to provide a concrete implementation of the ``Manager`` class and ``Destination``.
+
+1) Define a custom ``Serializer``, ``Manager``, ``Destination`` and optionally a subclass of ``TelemetryInfo``
+
+.. code-block:: c++
+
+  class JsonSerializer : public Serializer {
+  public:
+    json::Object *getOutputObject() { return object.get(); }
+
+    llvm::Error init() override {
+      if (started)
+        return createStringError("Serializer already in use");
+      started = true;
+      object = std::make_unique<json::Object>();
+      return Error::success();
+    }
+
+    // Serialize the given value.
+    void write(StringRef KeyName, bool Value) override {
+      writeHelper(KeyName, Value);
+    }
+
+    void write(StringRef KeyName, int Value) override {
+      writeHelper(KeyName, Value);
+    }
+
+    void write(StringRef KeyName, size_t Value) override {
+      writeHelper(KeyName, Value);
+    }
+    void write(StringRef KeyName, StringRef Value) override {
+      writeHelper(KeyName, Value);
+    }
+
+    void write(StringRef KeyName,
+               const std::map<std::string, std::string>& Value) override {
+      json::Object Inner;
+      for (auto kv : Value) {
+        Inner.try_emplace(kv.first, kv.second);
+      }
+      writeHelper(KeyName, json::Value(std::move(Inner)));
+    }
+
+    Error finalize() override {
+      if (!started)
+        return createStringError("Serializer not currently in use");
+      started = false;
+      return Error::success();
+    }
+
+  private:
+    template <typename T> void writeHelper(StringRef Name, T Value) {
+      assert(started && "serializer not started");
+      object->try_emplace(Name, Value);
+    }
+    bool started = false;
+    std::unique_ptr<json::Object> object;
+  };
+       
+  class MyManager : public telemery::Manager {
+  public:
+  static std::unique_ptr<MyManager> createInstatnce(telemetry::Config* config) {
+    // If Telemetry is not enabled, then just return null;
+    if (!config->EnableTelemetry) return nullptr;
+
+    return std::make_unique<MyManager>();
+  }
+  MyManager() = default;
+
+  Error dispatch(TelemetryInfo* Entry) const override {
+    Entry->SessionId = SessionId;
+    emitToAllDestinations(Entry);
+  }
+      
+  void addDestination(std::unique_ptr<Destination> dest) override {
```

Comment:
Should `dest` be an r-value reference? (Also `dest` -> `Dest`)

Reply:
I think we'd like the manager to own/manage the Destination objects for simplicity

Reply:
> Right, but passing `std::unique_ptr<Destination> &&` achieves that? I'm pretty sure it's more common to pass them like that.

Not sure about "more common" (tbh, it's VERY rare in our codebase) but passing it by reference does not communicate clearly the transfer of ownership, compared to passing-by-value. Also detriment to readability.

Reply:
I searched in LLVM before making that claim and found over 12000 instances of the regex `std::unique_ptr<.*> &&` in the codebase (when including clang and lld). It's possible I got my regex wrong, but a cursory glance verified it on the two or three I clicked on.

I also personally disagree about the readability. Technically you could std::move a unique_ptr unnecessarily, but why would you? I think the use of `std::move` at the callsite makes it clear that you are passing ownership. I'm less fussed about this argument though.

Reply:
In llvm-project repo: `git grep 'unique_ptr<.*> &&' | wc -l` -> 261?

The readability question is about definite ownership transfer, vs maybe-ownership-transfer. If you use `&&`, it is not as clear you have passed ownership -- you might well not be doing so!

That is: if you have a function, `void f(unique_ptr<int> x);`, and call the function like `std::unique_ptr<int> local_variable; ...; f(std::move(local_variable));`, then `local_variable.get()` will be definitely null after the call; ownership was definitely transferred to the function argument.

If the function is `void f(unique_ptr<int> &&x)`, and call it the same way, no ownership transfer has occurred as part of the call, so, `local_variable.get()` _may or may not_ be null after the call, depending on whether the body of `f` transfers ownership of `x` somewhere.

It is unusual to have a situation which actually requires the semantics of "might take ownership, but might not", so it is nearly always better to pass unique_ptr by value.

Reply:
Hehe, I'm dropping my concern. I _may_ have accidentally been looking at the "total files searched" value, rather than the number of cases found 🤦‍♂️